CORE-525: Handle Pending Transactions

EBGToo opened the pull request:
Currently draft as we are waiting on BDB to announce submitted transactions as 'SUBMITTED' (currently we get nothing until 'included' even thought the transaction is properly forwarded to the BTC peers and in their mempools).

michael-brd commented on bitcoin/BRSyncManager.c:
```diff
@@ -1162,25 +1164,23 @@ BRClientSyncManagerAnnounceGetTransactionsItem (BRClientSyncManager manager,
 
+    // Wallet already knows about this txn; so just update the block info.
+    if (/* isRegistered */ NULL != transaction && BRTransactionIsSigned (transaction)) {
+        BRWalletUpdateTransactions (manager->wallet, &transaction->txHash, 1, (uint32_t) blockHeight, (uint32_t) timestamp);
```
This should be in the above `needRegistration` if block, ya? (obviates the need to repeat the NULL and signed check, also respects the request ID being correct.

EBGToo replied:
Don't we want to update the transaction even if we don't need to register it?

EBGToo replied:
I updated the logic with a clear check for if the update should occur.

michael-brd commented on bitcoin/BRSyncManager.c:
```diff
@@ -1279,7 +1279,9 @@ BRClientSyncManagerAnnounceGetTransactionsDone (BRClientSyncManager manager,
                                                       addressCount,
                                                       begBlockNumber,
-                                                      endBlockNumber,
+                                                      (BRClientSyncManagerScanStateIsFullScan (&manager->scanState)
```
Can this behaviour be folded into `BRClientSyncManagerScanStateGetEndBlockNumber`? The idea with all those `BRClientSyncManagerScanState` functions was to encapsulate, as much as possible, the scan behaviour/logic.

michael-brd replied:
(alternatively, could set the `scanState->endBlockNumber` in `BRClientSyncManagerScanStateInit`, allowing `BRClientSyncManagerScanStateGetEndBlockNumber` to remain the same.

EBGToo replied:
If so, please go ahead and change as you see fit.

michael-brd replied:
Okay, give me 10.

michael-brd commented on bitcoin/BRSyncManager.c:
```diff
@@ -1353,7 +1355,9 @@ BRClientSyncManagerUpdateTransactions (BRClientSyncManager manager) {
                                                       addressCount,
                                                       begBlockNumber,
-                                                      endBlockNumber,
+                                                      (BRClientSyncManagerScanStateIsFullScan (&manager->scanState)
```
Same comment re:moving to `BRClientSyncManagerScanStateInit`/`BRClientSyncManagerScanStateGetEndBlockNumber`.

michael-brd commented on support/BRBase.h:
```diff
@@ -26,13 +26,34 @@
+ * Check if `height` is between the MINIMUM and MAXIMUM values.
+ */
+#define BLOCK_HEIGHT_IS_IN_RANGE(height) \
+AS_CRYPTO_BOOLEAN (BLOCK_HEIGHT_MINIMUM <= (height) && (height) <= BLOCK_HEIGHT_MAXIMUM)
```
Might lead to dependency issues here (`AS_CRYPTO_BOOLEAN` comes from crypto/BRCryptoBase.h). Maybe we move `boolean` into support sooner rather than later?

EBGToo replied:
I'll remove the `AS_CRYPTO_BOOLEAN` use.

michael-brd commented on bitcoin/BRSyncManager.c:
```diff
@@ -1162,25 +1164,23 @@ BRClientSyncManagerAnnounceGetTransactionsItem (BRClientSyncManager manager,
+            if (transaction == BRWalletTransactionForHash (manager->wallet, transaction->txHash)) {
                 // If our transaction made it into the wallet, do not deallocate it
-                transaction = NULL;
+                needFree = 0;
```
Nice.

michael-brd commented on Swift/BRCrypto/BRCryptoSystem.swift:
```diff
@@ -1196,8 +1196,8 @@ extension System {
                                                addresses: addresses,
-                                               begBlockNumber: begBlockNumber,
-                                               endBlockNumber: endBlockNumber,
+                                               begBlockNumber: (begBlockNumber == BLOCK_HEIGHT_UNBOUND ? nil : begBlockNumber),
```
What does a nil begin block number map to, a scan from 0 presumably?

EBGToo replied:
I don't know how BDB responds to a non-existent 'start_block', probably errors.

michael-brd replied:
> > @EBGToo, looks good. I updated the Java. I changed the Java wallet sweeper to use the `null` start/end height (rather than 0/`Network.height`). Does that seem like a reasonable change? Am I understanding correctly the BDB behaviour when the start number is omitted?
> 
> I don't know the BDB behavior if `start_height` is omitted. I made both 'start' and 'end' optional for uniformity reasons.

Okay, I walked back that change. Thanks!

michael-brd commented:
@EBGToo, looks good. I updated the Java. I changed the Java wallet sweeper to use the `null` start/end height (rather than 0/`Network.height`). Does that seem like a reasonable change? Am I understanding correctly the BDB behaviour when the start number is omitted?

EBGToo commented:
> @EBGToo, looks good. I updated the Java. I changed the Java wallet sweeper to use the `null` start/end height (rather than 0/`Network.height`). Does that seem like a reasonable change? Am I understanding correctly the BDB behaviour when the start number is omitted?

I don't know the BDB behavior if `start_height` is omitted.  I made both 'start' and 'end' optional for uniformity reasons.

michael-brd commented:
> > @EBGToo, looks good. I updated the Java. I changed the Java wallet sweeper to use the `null` start/end height (rather than 0/`Network.height`). Does that seem like a reasonable change? Am I understanding correctly the BDB behaviour when the start number is omitted?
> 
> I don't know the BDB behavior if `start_height` is omitted. I made both 'start' and 'end' optional for uniformity reasons.

Okay, I walked back that change. Thanks!

EBGToo commented:
> Should code be lowercased here?

We got the case correct with the later commits, it seems.
